src/analysis/ast_dependency_detector.rs

```diff
@@ -46,14 +48,105 @@ pub struct ASTDependencyDetector<'a> {
         )>,
     >,
     params: Option<Vec<TypedVar<'a>>>,
+    top_level: bool,
     preloaded: &'a BTreeMap<QualifiedContractIdentifier, ContractAST>,
 }
 
+#[derive(Clone, Debug, Eq)]
+pub struct Dependency {
+    pub contract_id: QualifiedContractIdentifier,
+    pub required_before_publish: bool,
+}
+
+impl PartialEq for Dependency {
+    fn eq(&self, other: &Self) -> bool {
+        self.contract_id == other.contract_id
```

Comment:
Shouldn't we including `required_before_publish` in this equality?

Reply:
See the `add_dependency` method. What I was going for is that a contract is only in the set once (regardless of this flag), and a dependency with the `required_before_publish` flag overwrites a dependency without that flag. Is this a confusing way to do that?

Reply:
Works for me, thank you unpacking!